libs/sdk-ui-web-components/src/visualizations/CustomElementAdapter.ts

```diff
@@ -0,0 +1,173 @@
+// (C) 2022 GoodData Corporation
+import React from "react";
+import ReactDom from "react-dom";
+import invariant from "ts-invariant";
+import { IAnalyticalBackend } from "@gooddata/sdk-backend-spi";
+import { LoadingComponent } from "@gooddata/sdk-ui/esm/base/react/LoadingComponent";
+
+import { CustomElementContext, getContext } from "../context";
+// eslint-disable-next-line import/no-unassigned-import
+import "./components.shadow.css";
+// eslint-disable-next-line import/no-unassigned-import
+import "./components.css";
+
+// Since JS does not support private properties natively,
+// use Symbols to hide them. Otherwise, would be seen on the element instance.
+const MOUNT_POINT = Symbol("mountPoint");
+const RENDER = Symbol("render");
+const COMPONENT = Symbol("component");
+const CONTEXT = Symbol("context");
+export const GET_VISUALIZATION = Symbol("getVisualization");
+export const EVENT_HANDLER = Symbol("eventHandler");
+export const LOAD_COMPONENT = Symbol("loadComponent");
+
+export abstract class CustomElementAdapter<C> extends HTMLElement {
+    /**
+     * @remarks
+     * A mounting point for the React visualization
+     *
+     * @internal
+     */
+    private [MOUNT_POINT]: HTMLDivElement;
+
+    /**
+     * @remarks
+     * A React Component to be used to render the visualization
+     *
+     * @internal
+     */
+    private [COMPONENT]: C;
+
+    /**
+     * @remarks
+     * The context for the visualization rendering (backend, workspace etc.)
+     *
+     * @internal
+     */
+    private [CONTEXT]: CustomElementContext;
+
+    constructor() {
+        super();
+
+        // Attach Shadow DOM
+        const shadowRoot = this.attachShadow({ mode: "closed" });
+
+        // Inject styles to the node
+        const link = document.createElement("link");
+        link.type = "text/css";
+        link.rel = "stylesheet";
+        const webpackWorkaround = SHADOW_STYLES;
+        link.href = new URL(webpackWorkaround, import.meta.url).href;
```

Comment:
There is a weird bug with Webpack when it's trying to interpolate `new URL('/some/path', import.meta.url)` build time. I guess it's an optimization of some sort, as they are using this pattern internally as well. Re-assigning the string value to a variable first works around that issue.

Reply:
Note, `SHADOW_STYLES` is interpolated to a string value by the `DefinePlugin`. See webpack config.